File: Linked List/singly_linked_list.py
'''
This is an implementation of a singly linked list in Python with a Node class and a LinkedList class. The Node class represents each element of the linked list and the LinkedList class has methods to manipulate the list, including append, prepend, delete_node, and print_list.
'''

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None  # The next node in the list

class LinkedList:
    def __init__(self):
        self.head = None  # The first node in the list

    def append(self, data):
        new_node = Node(data)
        if not self.head:  # If the list is empty, set the new node as the head
            self.head = new_node
            return
        curr_node = self.head
        while curr_node.next:  # Traverse to the last node in the list
            curr_node = curr_node.next
        curr_node.next = new_node  # Set the next node of the last node to the new node

    def prepend(self, data):
        new_node = Node(data)
        new_node.next = self.head  # Set the next node of the new node to the current head
        self.head = new_node  # Set the new node as the new head

    def delete_node(self, data):
        if not self.head:  # If the list is empty, do nothing
            return
        if self.head.data == data:  # If the head is the node to delete, set the next node as the new head
            self.head = self.head.next
            return
        curr_node = self.head
        while curr_node.next:  # Traverse the list until the last node
            if curr_node.next.data == data:  # If the next node is the node to delete, set the next node of the current node to the node after the next node
                curr_node.next = curr_node.next.next
                return
            curr_node = curr_node.next

    def print_list(self):
        curr_node = self.head
        while curr_node:  # Traverse the list and print each node's data
            print(curr_node.data)
            curr_node = curr_node.next

File: Linked List/singly_linked_list.java
// Linked list: Implement Singly linked list in Java #982

/*
APPROACH :Node Class: The Node class is defined to represent a single node in the linked list. It has two instance variables: data to store the value and next to store the reference to the next node.

Insertion at the Start: The insertStart method allows inserting a new node at the beginning of the linked list. It creates a new node with the given data value and sets its next reference to the current head. Then, it updates the head to point to the newly inserted node.

Deletion from the Start: The delete method removes the first node from the linked list. It checks if the head is null, indicating an empty list. If not, it updates the head to point to the next node, effectively removing the first node.

Display: The display method is used to print the elements of the linked list. It starts from the head node and iterates through the list by moving to the next node until the current node becomes null. During each iteration, it prints the data value of the current node.

Main Method: The main method demonstrates the usage of the linked list implementation. It creates an empty list by initializing the head to null. It then performs several insertions using the insertStart method to add nodes at the beginning of the list. After that, it calls the display method to print the elements of the list. Next, it performs several deletions using the delete method to remove nodes from the beginning of the list. Finally, it calls the display method again to print the updated list.
*/

import java.lang.*;

// Node Class
class Node {
    int data;
    Node next;

    Node(int x) // parameterized constructor
    {
        data = x;
        next = null;
    }
}

class Main
{
    static Node insertStart(Node head, int data)
    {
        // Creating newNode memory & assigning data value
        Node newNode = new Node(data);

        // assigning this newNode's next as current head node
        newNode.next = head;
        // re-assigning head to this newNode
        head = newNode;

        return head;
    }

    public static Node delete(Node head)
    {
        if (head == null){
            System.out.println("List is empty, not possible to delete");
            return head;
        }

        System.out.println("Deleted: " + head.data);
        // move head to next node
        head = head.next;

        return head;
    }

    static void display(Node node) {

        //as linked list will end when Node is Null
        while (node != null) {
            System.out.print(node.data + " ");
            node = node.next;
        }
        System.out.println("");
    }

    public static void main(String args[])
    {
        Node head = null;
        head = insertStart(head,6);
        head = insertStart(head,5);
        head = insertStart(head,4);
        head = insertStart(head,3);
        head = insertStart(head,2);
        head = insertStart(head,1);

        display(head);

        head = delete(head);
        head = delete(head);
        head = delete(head);

        display(head);


    }
}

/*
the space complexity is O(n), and the time complexity for each operation is either O(1) or O(n), depending on the specific operation being performed.
*/
